File: src/sudoku.cc


#include "sudoku.h"


using std::cout;
using std::endl;
using std::ostream;
using std::vector;


inline int index(int8_t* array, int row, int col) {
  return (int) array[row*BOARD_COLS + col];
}


inline void setindex(int8_t* array, int row, int col, int8_t value) {
  array[row*BOARD_COLS + col] = value;
}


inline int8_t* index(int8_t** array, int row, int col) {
  return array[row*BOARD_COLS + col];
}


Sudoku::Sudoku() {
  allocate();
}


Sudoku::~Sudoku() {
  free();
}


void Sudoku::allocate() {
  answer = new int8_t[BOARD_SIZE];
  for(int i = 0; i < BOARD_SIZE; i++) {
    answer[i] = 0;
  }

  sizes = new int8_t[BOARD_SIZE];
  for(int i = 0; i < BOARD_SIZE; i++) {
    sizes[i] = 9;
  }

  available = new int8_t*[BOARD_SIZE];
  for(int i = 0; i < BOARD_SIZE; i++) {
    available[i] = new int8_t[9];

    for(int j = 0; j < 9; j++) {
      available[i][j] = (int8_t) j+1;
    }
  }
}


void Sudoku::free() {
  delete [] answer;
  delete [] sizes;

  for(int i = 0; i < BOARD_SIZE; i++) {
    if(available[i] != NULL) {
      delete [] available[i];
    }
  }
  delete [] available;
}


ostream& operator<<(ostream& os, const Sudoku& s) {
  // Print the available values
  cout << " ----------------------------------------------------------";
  cout << "---------------------------------------" << endl;
  cout << "                                             AVAILABLE     ";
  cout << "                                       " << endl;
  for(int i = 0; i < BOARD_ROWS; i++) {
    if(i % 3 == 0) {
      cout << " ----------------------------------------------------------";
      cout << "---------------------------------------" << endl;
    }

    for(int j = 0; j < BOARD_COLS; j++) {
      if(j%3 == 0)
        cout << " |";

      cout << " ";

      int idx = 0;
      for(int k = 0; k < 9; k++) {
        if(k+1 == index(s.available, i, j)[idx]) {
          cout << k+1;
          
          if(idx < index(s.sizes, i, j)-1)
            idx++;
        } else {
          cout << " ";
        }
      }
    }

    cout << " |" << endl;
  }
  cout << " ----------------------------------------------------------";
  cout << "---------------------------------------" << endl;

  // Print the currently answered values
  cout << " -------------------------" << endl;
  cout << "           ANSWER         " << endl;
  for(int i = 0; i < BOARD_ROWS; i++) {
    if(i % 3 == 0)
      cout << " -------------------------" << endl;

    for(int j = 0; j < BOARD_COLS; j++) {
      if(j%3 == 0)
        cout << " |";

      cout << " " << index(s.answer, i, j);
    }

    cout << " |" << endl;
  }
  cout << " -------------------------" << endl;

  return os;
}


void Sudoku::addCage(int8_t sum, const vector<Cell>& cells) {
  if(cells.size() == 1) {
    setSolvedCell(cells[0].x, cells[0].y, sum);
  } else {
    Cage cage;
    cage.sum = sum;
    cage.cells = cells;
    cages.push_back(cage);
  }
}


void Sudoku::setSolvedCell(int x, int y, int8_t value) {
  setindex(answer, y, x, value);
  clearCol(x, y, value);
  clearRow(x, y, value);
  clearCage(getCage(x, y), value);

  int idx = y*BOARD_COLS + x;
  delete [] available[idx];
  available[idx] = new int8_t[1];
  available[idx][0] = value;
  sizes[idx] = 1;
}


void Sudoku::clearCol(int col, int row, int8_t value) {
  for(int i = 0; i < BOARD_ROWS; i++) {
    removeAvailable(col, i, value);
  }
}


void Sudoku::clearRow(int col, int row, int8_t value) {
  for(int i = 0; i < BOARD_COLS; i++) {
    removeAvailable(i, row, value);
  }
}


void Sudoku::clearCage(int n, int8_t value) {
  cout << "clear cage " << n << ": " << (int) value << endl;
}


int Sudoku::getCage(int x, int y) {
  int row = x / 3;
  int col = y / 3;
  return row * BOARD_ROWS/3 + col;
}


void Sudoku::removeAvailable(int col, int row, int8_t value) {
  int size = index(sizes, row, col);
  for(int i = 0; i < size; i++) {
    if(index(available, row, col)[i] == value) {
      int8_t* modified = new int8_t[size-1];
      int idx = row*BOARD_COLS + col;

      for(int j = 0; j < size; j++) {
        if(j < i)
          modified[j] = available[idx][j];
        else if(j > i)
          modified[j-1] = available[idx][j];
      }

      delete [] available[idx];
      available[idx] = modified;
      sizes[idx] = size-1;
      return;
    }
  }

  if(size == 1) {
    setSolvedCell(col, row, value);
  }
}


bool Sudoku::checkSolved() {
  for(int i = 0; i < BOARD_SIZE; i++) {
    if(answer[i] == 0)
      return false;
  }
  return true;
}


void Sudoku::solve() {
  while(1) {
    if(checkSolved()) break;
    if(killerCombinations()) continue;

    cout << "Cannot solve this puzzle! We run out of strategies!" << endl;
    break;
  } 
}

File: domain/sudoku.go
package domain

import (
	"fmt"
	"os"
)

const (
	BOARD_COLS = 9
	BOARD_ROWS = 9
	BOARD_SIZE = BOARD_COLS * BOARD_ROWS
)

const (
	AVAILABLE_1 = 1 << iota
	AVAILABLE_2
	AVAILABLE_3
	AVAILABLE_4
	AVAILABLE_5
	AVAILABLE_6
	AVAILABLE_7
	AVAILABLE_8
	AVAILABLE_9

	AVAILABLE_ALL = AVAILABLE_1 | AVAILABLE_2 | AVAILABLE_3 | AVAILABLE_4 |
		AVAILABLE_5 | AVAILABLE_6 | AVAILABLE_7 | AVAILABLE_8 | AVAILABLE_9
)

type Sudoku struct {
	Cages     []*Cage
	Answer    []int8
	Available []int
}

func NewSudoku() *Sudoku {
	available := make([]int, BOARD_SIZE)
	for i, _ := range available {
		available[i] = AVAILABLE_ALL
	}

	return &Sudoku{
		Answer:    make([]int8, BOARD_SIZE),
		Available: available,
	}
}

func (s *Sudoku) ReadFrom(filename string) error {
	f, err := os.Open(filename)
	if err != nil {
		return NewError(err)
	}
	defer f.Close()

	if err := s.readCages(f); err != nil {
		return err
	}

	return nil
}

func (s *Sudoku) Print() {
	fmt.Print(" -----------------------------------------------------")
	fmt.Println("--------------------------------------------")
	fmt.Println("                                             AVAILABLE")

	for i := 0; i < BOARD_ROWS; i++ {
		if i%3 == 0 {
			fmt.Print(" -----------------------------------------------------")
			fmt.Println("--------------------------------------------")
		}

		for j := 0; j < BOARD_COLS; j++ {
			if j%3 == 0 {
				fmt.Print(" | ")
			} else {
				fmt.Print("-")
			}

			idx := i*BOARD_COLS + j

			var n uint
			var k int8
			for ; k < 9; k++ {
				mask := 1 << n
				if s.Available[idx]&mask == mask {
					fmt.Print(k + 1)
					n++
				} else {
					fmt.Print(" ")
				}
			}
		}
		fmt.Println(" |")
	}

	fmt.Print(" -----------------------------------------------------")
	fmt.Println("--------------------------------------------")

	fmt.Println(" -------------------------")
	fmt.Println("          ANSWER")

	for i := 0; i < BOARD_ROWS; i++ {
		if i%3 == 0 {
			fmt.Println(" -------------------------")
		}

		for j := 0; j < BOARD_COLS; j++ {
			if j%3 == 0 {
				fmt.Print(" | ")
			} else {
				fmt.Print(" ")
			}

			n := s.Answer[i*BOARD_COLS+j]
			if n == 0 {
				fmt.Print(" ", "")
			} else {
				fmt.Print(n, "")
			}
		}
		fmt.Println(" |")
	}

	fmt.Println(" -------------------------")
}

/*
func (s *Sudoku) SetAvailable(row, col int8, av []int8) {
	s.Available[row*BOARD_COLS+col] = make([]int8, len(av))
	copy(s.Available[row*BOARD_COLS+col], av)
}

func (s *Sudoku) RemoveAvailable(row, col int, value int8) error {
	idx := row*BOARD_COLS + col

	for i, n := range s.Available[idx] {
		if n == value {
			if len(s.Available[idx]) == 1 {
				return NewErrorf("cannot remove the last available number: %dx%d", row, col)
			}

			s.Available[idx] = append(s.Available[idx][:i], s.Available[idx][i+1:]...)
			return nil
		}
	}

	return nil
}

func (s *Sudoku) Solved() bool {
	for _, c := range s.Answer {
		if c == 0 {
			return false
		}
	}
	return true
}

func (s *Sudoku) SolvedSquares() (modified bool, e error) {
	for i, available := range s.Available {
		if s.Answer[i] == 0 && len(available) == 1 {
			fmt.Printf(" * Found solved square: %dx%d: %d\n", i/BOARD_COLS,
				i%BOARD_COLS, available[0])

			if err := s.SolveCell(i/BOARD_COLS, i%BOARD_COLS, available[0]); err != nil {
				return false, err
			}

			modified = true
		}
	}

	return modified, nil
}

func (s *Sudoku) SolveCell(row, col int, sol int8) error {
	idx := row*BOARD_COLS + col

	s.Answer[idx] = sol
	if len(s.Available[idx]) != 1 {
		s.Available[idx] = []int8{sol}
	}

	// Clear the row
	for i := 0; i < BOARD_ROWS; i++ {
		if i == row {
			continue
		}

		if err := s.RemoveAvailable(i, col, sol); err != nil {
			return err
		}
	}

	// Clear the col
	for i := 0; i < BOARD_COLS; i++ {
		if i == col {
			continue
		}

		if err := s.RemoveAvailable(row, i, sol); err != nil {
			return err
		}
	}

	// Clear the cage
	var x int = col / 3
	x *= 3
	var y int = row / 3
	y *= 3

	for i := x; i < x+3; i++ {
		for j := y; j < y+3; j++ {
			if i == col && j == row {
				continue
			}

			if err := s.RemoveAvailable(j, i, sol); err != nil {
				return err
			}
		}
	}

	return nil
}
*/
func (s *Sudoku) readCages(f *os.File) error {
	var ncages int
	fmt.Fscan(f, &ncages)

	s.Cages = make([]*Cage, ncages)

	for i := 0; i < ncages; i++ {
		var sum int8
		var ncells int
		fmt.Fscan(f, &sum, &ncells)

		cage := &Cage{
			Sum:   sum,
			Cells: make([]*Cell, ncells),
		}

		var row, col int8
		for j := 0; j < ncells; j++ {
			fmt.Fscanf(f, "%d,%d", &col, &row)
			cage.Cells[j] = &Cell{
				Row: row,
				Col: col,
			}
		}

		s.Cages[i] = cage
	}

	return nil
}
